librosa/core/notation.py

```diff
@@ -0,0 +1,509 @@
+#!/usr/bin/env python
+# -*- coding: utf-8 -*-
+'''Music notation utilties'''
+
+import re
+import numpy as np
+from .._cache import cache
+from ..util.exceptions import ParameterError
+
+__all__ = ['key_to_degrees', 'key_to_notes',
+           'mela_to_degrees', 'mela_to_svara',
+           'thaat_to_degrees']
+
+THAAT_MAP = dict(bilaval    =    [0, 2, 4, 5, 7, 9, 11],
+                 khamaj     =    [0, 2, 4, 5, 7, 9, 10],
+                 kafi       =    [0, 2, 3, 5, 7, 9, 10],
+                 asavari    =    [0, 2, 3, 5, 7, 8, 10],
+                 bhairavi   =    [0, 1, 3, 5, 7, 8, 10],
+                 kalyan     =    [0, 2, 4, 6, 7, 9, 11],
+                 marva      =    [0, 1, 4, 6, 7, 9, 11],
+                 poorvi     =    [0, 1, 4, 6, 7, 8, 11],
+                 todi       =    [0, 1, 3, 6, 7, 8, 11],
+                 bhairav    =    [0, 1, 4, 5, 7, 8, 11])
+
+MELAKARTA_MAP = {k: i
+                 for i, k in enumerate(['kanakanki', 'ratnangi', 'ganamurti',
+                                        'vanaspati', 'manavati', 'tanarupi',
+                                        'senavati', 'hanumatodi', 'dhenuka',
+                                        'natakapriya', 'kokilapriya', 'rupavati',
+                                        'gayakapriya', 'vakulabharanam', 'mayamalavagoulai',
+                                        'chakravaham', 'suryakantam', 'hatakambhari',
+                                        'jhankaradhwani', 'natabhairavi', 'keeravani',
+                                        'kharaharapriya', 'gowrimanohari', 'varunapriya',
+                                        'mararanjani', 'charukesi', 'sarasangi',
+                                        'harikambhoji', 'dheerasankarabharanam', 'naganandini',
+                                        'yagapriya', 'ragavardhini', 'gangeyabhusani',
+                                        'vagadheeswari', 'sulini', 'chalanattai',
+                                        'salagam', 'jalarnavam', 'jhalavarali',
+                                        'navaneetam', 'pavani', 'raghupriya',
+                                        'gavambodhi', 'bhavapriya', 'subhapantuvarali',
+                                        'shadvigamargini', 'suvarnangi', 'divyamani',
+                                        'dhavalambari', 'namanarayani', 'kamavardhini',
+                                        'ramapriya', 'gamanasrama', 'viswambhari',
+                                        'syamalangi', 'shanmukhapriya', 'simhendramadhyamam',
+                                        'hemavati', 'dharmavati', 'nitimati',
+                                        'kantamani', 'rishabhapriya', 'latangi',
+                                        'vachaspati', 'mechakalyani', 'chitrambhari',
+                                        'sucharitra', 'jyotiswarupini', 'dhatuvardhini',
+                                        'nasikabhushani', 'kasalam', 'rasikapriya'])}
+
```

Comment:
Hi @bmcfee,
As you have already noted, there will be inconsistencies in how melakarta names (originally in Sanskrit) is written in English. I have suggested a few modifications to the English spellings which closely matches their Sanskrit pronunciations. These suggestions are based on me being Carnatic music student for ~15 years (& learnt Sanskrit in school) and married to a person who has been a Carnatic music teacher for the last decade.

I would prefer 1-based indexing in melakarta over 0-based. Also, I have generally seen "svara" written as "swara". Sorry for being pedantic but, please feel free to reject any/all suggestions.

['kanakaangi', 'ratnaangi', 'gaanamurti',
'vanaspati', 'maanavati', 'taanarupi',
'senaavati', 'hanumatodi', 'dhenuka',
'natakapriya', 'kokilapriya', 'rupavati',
'gayakapriya', 'vakulaabharanam', 'mayaamaalavagowla',
'chakravaakam', 'suryakantam', 'hatakaambhari',
'jhankaaradhwani', 'natabhairavi', 'keeravani',
'kharaharapriya', 'gowrimanohari', 'varunapriya',
'maararanjani', 'chaarukesi', 'sarasaangi',
'harikambhoji', 'dheerasankarabharanam', 'naganandini',
'yaagapriya', 'raagavardhini', 'gaangeyabhushani',
'vagadheeswari', 'shulini', 'chalanaatta',
'saalagam', 'jalaarnavam', 'jhaalavaraali',
'navaneetam', 'paavani', 'raghupriya',
'gavaambodhi', 'bhavapriya', 'subhapanthuvaraali',
'shadvigamargini', 'suvarnaangi', 'divyamani',
'dhavalambari', 'naamanarayani', 'kaamavardhini',
'raamapriya', 'gamanasrama', 'viswambhari',
'syaamalaangi', 'shanmukhapriya', 'simhendramadhyamam',
'hemavati', 'dharmavati', 'neetimati',
'kaantaamani', 'rishabhapriya', 'lataangi',
'vachaspati', 'mechakalyani', 'chitrambhari',
'sucharitra', 'jyotiswarupini', 'dhatuvardhini',
'nasikabhushani', 'kosalam', 'rasikapriya']

Reply:
Pedantic is the name of the game here, so thanks for this!


> I would prefer 1-based indexing in melakarta over 0-based.

Gotcha.  If I can push on that a little more, would it be common to refer by number like this?  IE if I said `melakarta #5`, knowledgeable people would know what is meant?

> Also, I have generally seen "svara" written as "swara".

I've been seeing both in relatively equal quantities.  For example, the [wikipedia article](https://en.wikipedia.org/wiki/Svara) uses `Svara` as the title, and `swara` and `svara` interchangeably throughout the body.  (Not that wikiepedia is a canonical reference here, but I expect it is at least somewhat representative of diversity of interested in parties.  I see similar divergence across different reference texts, so I'm really not sure here.)

I imagine this comes from deviations in regional dialect and maybe inconsistent conventions for translation orthography, and I don't want to unduly privilege one translation over another without good reason.  This [quora post](https://www.quora.com/Does-Hindi-have-an-official-romanized-alphabet-like-Pinyin-for-Mandarin) links out to [IAST](https://en.wikipedia.org/wiki/International_Alphabet_of_Sanskrit_Transliteration), but this also doesn't seem to be of much help here.




> I have suggested a few modifications to the English spellings which closely matches their Sanskrit pronunciations.

It's difficult for me to eyeball this and quickly get a sense of the changes you're suggesting.  Could you briefly explain the thinking behind them, or at least a couple of examples?

Reply:
> Pedantic is the name of the game here, so thanks for this!
> 
Thanks :-) 

> > I would prefer 1-based indexing in melakarta over 0-based.
> 
> Gotcha. If I can push on that a little more, would it be common to refer by number like this? IE if I said `melakarta #5`, knowledgeable people would know what is meant?

Absolutely! 
In fact, the melakarta number can be deduced from the name of each raga.  The first two-syllable of the Raga name has a specific number associated with it based on the ancient [ _Katapayadi_ system](https://artofmemory.com/wiki/Katapayadi_System). For example, the first two-syllable of the raga dheerasankarabharanam:  _dhee_  has the number 9 associated with it and _ra_ has the number 2 associated with it. Hence, the first two syllables of this raga represent the number 92. However, according to the _Katapayaadi_ _Sankya_ the syllable _ra_ appears before the syllable _dhee_ hence, the numbers are to be reversed and you get 29 which is the number associated with deerasankarabharanam on the melakartha system. Going further, given a number of melakarta, it is possible to deduce all the svaras (or swaras) that should exist in that raga along with the first two syllables of the name. Hence, 1-based indexing makes absolute sense to a trained musician while 0-based means nothing.

> 
> > Also, I have generally seen "svara" written as "swara".
> 
> I've been seeing both in relatively equal quantities. For example, the [wikipedia article](https://en.wikipedia.org/wiki/Svara) uses `Svara` as the title, and `swara` and `svara` interchangeably throughout the body. (Not that wikiepedia is a canonical reference here, but I expect it is at least somewhat representative of diversity of interested in parties. I see similar divergence across different reference texts, so I'm really not sure here.)
> 
> I imagine this comes from deviations in regional dialect and maybe inconsistent conventions for translation orthography, and I don't want to unduly privilege one translation over another without good reason. This [quora post](https://www.quora.com/Does-Hindi-have-an-official-romanized-alphabet-like-Pinyin-for-Mandarin) links out to [IAST](https://en.wikipedia.org/wiki/International_Alphabet_of_Sanskrit_Transliteration), but this also doesn't seem to be of much help here.

Agreed! It is always a challenge to translate from Sanskrit, which has more letters (and different grammar) to English. I can live with 'svara'! 👍 :-)

> 
> > I have suggested a few modifications to the English spellings which closely matches their Sanskrit pronunciations.
> 
> It's difficult for me to eyeball this and quickly get a sense of the changes you're suggesting. Could you briefly explain the thinking behind them, or at least a couple of examples?

The changes are mostly to do with what syllables should have _extended pronounciations_. Sorry, I don't know how exactly to convey this in the text (I wish I could record my voice and send!). Here is an analogy: the word _marquee_ has an extended pronunciation of the _e_. This wouldn't be captured if I were to write it as _marki_ (but, English is a funny language and we don't always pronounce like the word is written). 
Example: ganamurti -> gaanamurti; Taking the first two-syllable; the _ah_ of _ga_ syllable is pronounced for slightly more duration than the _ah_ of _na_ syllable. Hence, writing _ganamurti_ as _gaanamurthi_ would indicate it. However, strictly speaking, it should be written as [gānamūrti](https://en.wikipedia.org/wiki/Ganamurti) but, it is not ASCII and hence I didn't follow those conventions of writing these names. 

Some of the names were wrong. kasalam -> kosalam

Sorry about the extended response. But, I am so glad that you are putting such great effort to get these things right and I feel compelled to do my part right as well (the least I could at this moment).

Please feel free to question / disagree and I will be happy to discuss these things further.

P.S.: I actually consulted my wife (who I consider a better musician than myself) about the spelling of some of my suggestions and we disagreed on a couple of them! That's mostly because we both have different native languages and our conventions of writing Sanskrit words in English differ! In short, it's impossible to get it "right" but, we can tend towards it!

Reply:
> Hence, 1-based indexing makes absolute sense to a trained musician while 0-based means nothing.

Ok, that's an easy enough change.  0-based makes the math a little simpler, but it's easy to adapt.  I'll push that change shortly.



> I can live with 'svara'!

I guess I'm more worried about, eg, if one might be more commonly used for Carnatic and the other for Hindustani (similar to re/ri).  I don't personally have a preference on this one.


>  However, strictly speaking, it should be written as [gānamūrti](https://en.wikipedia.org/wiki/Ganamurti) but, it is not ASCII and hence I didn't follow those conventions of writing these names.

Well, we *could* go unicode with this (since we're now in python 3 territory exclusively, this is pretty easy).  But I worry that it might not be so user-friendly that way.

The take-home message that I'm getting here is that your suggested corrections are there to emphasize stressed syllables, and that makes sense.  Still, I'd feel better about this if we could point to a reference (ideally text/book, but other reliable resources could be appropriate) rather than cobble it together piece-meal.

Also, it sounds like whatever we do, a helper function to list out our string names will be necessary.

> Some of the names were wrong. kasalam -> kosalam

Good catch.

Reply:
> The take-home message that I'm getting here is that your suggested corrections are there to emphasize stressed syllables, and that makes sense. Still, I'd feel better about this if we could point to a reference (ideally text/book, but other reliable resources could be appropriate) rather than cobble it together piece-meal.
> 

Please give me a couple of days and I will identify a textbook/ reference book that we could follow for the spelling of ragas.

> Also, it sounds like whatever we do, a helper function to list out our string names will be necessary.

It will be necessary, as some of the ragas are more commonly known in their shortened form (dheerasankarabharanam-> sankarabharanam; hanumatodi -> todi) or has an entirely different non-melakarta name (kamavardhini->panthuvarali). 

Would you be happy if I can put this information together? Please let me know.

Reply:
> Would you be happy if I can put this information together? Please let me know.

That'd be great, thanks!

Reply:
I did a little more digging through textbooks, and (unsurprisingly) came up with yet more disagreements.

The original transcriptions that I implemented in this PR came from "A gentle introduction to Carnatic Music" by Mahadevan Ramesh (2009) (`MR` below).

I also dug up an older reference text: "Ragas in Carnatic Music" by  S. Bhagyalekshmy (1990) (`SB` below).

Here's a quick listing of the two transcriptions, alongside @hskaushik 's (`KH` below) proposed corrections (only showing where they disagree):

|    | MR               | KH                 | SB                |
|---:|:-----------------|:-------------------|:------------------|
|  1 | kanakanki        | kanakaangi         | kanakangi         |
|  2 | ratnangi         | ratnaangi          | ratnangi          |
|  3 | ganamurti        | gaanamurti         | ganamurthi        |
|  4 | vanaspati        | vanaspati          | vanaspathi        |
|  5 | manavati         | maanavati          | manavathi         |
|  6 | tanarupi         | taanarupi          | tanarupi          |
|  7 | senavati         | senaavati          | senavathi         |
|  8 | hanumatodi       | hanumatodi         | hanumathodi       |
| 12 | rupavati         | rupavati           | rupavathi         |
| 14 | vakulabharanam   | vakulaabharanam    | vakulabharanam    |
| 15 | mayamalavagoulai | mayaamaalavagowla  | mayamalavagaula   |
| 16 | chakravaham      | chakravaakam       | chakravakom       |
| 17 | suryakantam      | suryakantam        | suryakantham      |
| 18 | hatakambhari     | hatakaambhari      | hatakambari       |
| 19 | jhankaradhwani   | jhankaaradhwani    | jhankaradhwani    |
| 23 | gowrimanohari    | gowrimanohari      | gaurimanohari     |
| 25 | mararanjani      | maararanjani       | mararanjini       |
| 26 | charukesi        | chaarukesi         | charukesi         |
| 27 | sarasangi        | sarasaangi         | sarasangi         |
| 28 | harikambhoji     | harikambhoji       | harukambhoji      |
| 31 | yagapriya        | yaagapriya         | yagapriya         |
| 32 | ragavardhini     | raagavardhini      | ragavardhini      |
| 33 | gangeyabhusani   | gaangeyabhushani   | gangeyabhushani   |
| 35 | sulini           | shulini            | sulini            |
| 36 | chalanattai      | chalanaatta        | chalanatta        |
| 37 | salagam          | saalagam           | salagam           |
| 38 | jalarnavam       | jalaarnavam        | jalarnavam        |
| 39 | jhalavarali      | jhaalavaraali      | jhalavarali       |
| 40 | navaneetam       | navaneetam         | navaneetham       |
| 41 | pavani           | paavani            | pavani            |
| 43 | gavambodhi       | gavaambodhi        | gavambodhi        |
| 45 | subhapantuvarali | subhapanthuvaraali | sudhapanthuvarali |
| 46 | shadvigamargini  | shadvigamargini    | shadvidhamargini  |
| 47 | suvarnangi       | suvarnaangi        | suvarnangi        |
| 50 | namanarayani     | naamanarayani      | namanarayani      |
| 51 | kamavardhini     | kaamavardhini      | kamavardhini      |
| 52 | ramapriya        | raamapriya         | ramapriya         |
| 55 | syamalangi       | syaamalaangi       | syamalangi        |
| 58 | hemavati         | hemavati           | hemavathi         |
| 59 | dharmavati       | dharmavati         | dharmavathi       |
| 60 | nitimati         | neetimati          | neethimathi       |
| 61 | kantamani        | kaantaamani        | kanthamani        |
| 63 | latangi          | lataangi           | latangi           |
| 64 | vachaspati       | vachaspati         | vachaspathi       |
| 66 | chitrambhari     | chitrambhari       | chitrambari       |
| 68 | jyotiswarupini   | jyotiswarupini     | jyotisvarupini    |

Differences I can spot:

1. KH's suggestions emphasize stress since we probably don't want to use diacritical marks and extended unicode here.
2. MR and SB differ primarily in `t` vs `th`.  26 differences in total, about 12 that are more substantial than t/th.

I expect that we'll see a similar number of differences between every pair of texts we compare.  I think our best option here is, if possible, to pick a suitably popular and widely available reference text and go with that, even if there will be disagreements.  However, I am certainly not qualified to judge which (if any) text is appropriate for this, so input from the experts would be much appreciated.

Reply:
I gave this another look through, and found some typos (my fault) in my transcription of SB.  (Going off a not-so-great scan of the text.)

Having corrected those, the disagreements to MR are reduced to the following:

|    | MR               | SB                |
|---:|:-----------------|:------------------|
|  1 | kanakanki        | kanakangi         |
|  3 | ganamurti        | ganamurthi        |
|  4 | vanaspati        | vanaspathi        |
|  5 | manavati         | manavathi         |
|  7 | senavati         | senavathi         |
|  8 | hanumatodi       | hanumathodi       |
| 12 | rupavati         | rupavathi         |
| 15 | mayamalavagoulai | mayamalavagaula   |
| 16 | chakravaham      | chakravakom       |
| 17 | suryakantam      | suryakantham      |
| 18 | hatakambhari     | hatakambari       |
| 23 | gowrimanohari    | gaurimanohari     |
| 25 | mararanjani      | mararanjini       |
| 33 | gangeyabhusani   | gangeyabhushani   |
| 36 | chalanattai      | chalanatta        |
| 40 | navaneetam       | navaneetham       |
| 45 | subhapantuvarali | sudhapanthuvarali |
| 46 | shadvigamargini  | shadvidhamargini  |
| 58 | hemavati         | hemavathi         |
| 59 | dharmavati       | dharmavathi       |
| 60 | nitimati         | neethimathi       |
| 61 | kantamani        | kanthamani        |
| 64 | vachaspati       | vachaspathi       |
| 66 | chitrambhari     | chitrambari       |
| 68 | jyotiswarupini   | jyotisvarupini    |

Aside from vowel elongations, SB looks pretty close to KH's suggestions. The one big oddball is 45, where "bha" (the apparently more popular choice) is "dha" in SB.  I'm hesitant to "correct" this, but it does seem unusual (if not erroneous).  

Apart from that, I think it makes sense to adopt SB's translations across the board, and otherwise not worry about it.

Reply:
> Aside from vowel elongations, SB looks pretty close to KH's suggestions. The one big oddball is 45, where "bha" (the apparently more popular choice) is "dha" in SB. I'm hesitant to "correct" this, but it does seem unusual (if not erroneous).

A quick note: I am certain the "dha" in SB should be typo! It should be "bha". Except that, I too am happy with SB over MR's version.

P.S: The Government of Karnataka (one of the southern states of India) conducts several grades (equivalent to undergraduate and Masters) carnatic music exams. The government publishes text books for these music courses and I am trying to source it through some contacts (unfortunately, the English version of this book is not available online). Hopefully, this could be our reference for Melakarta, when I manage to get it?

Reply:
> A quick note: I am certain the "dha" in SB should be typo! It should be "bha". Except that, I too am happy with SB over MR's version.

Thanks for confirming! :grin: That was my sense as well.

I've now pushed up the changes to SB's versions, so I think this should be good to go.



> Hopefully, this could be our reference for Melakarta, when I manage to get it?

That sounds like a great resource.  Not to be hasty, but I do want to get this released ASAP (this week, ideally), so I don't want to wait on additional references, nor do I want to put pressure on you to dig those up for me.

That said, I think we have a pretty good setup at this point.  One nice thing about the way that it's implemented is that we can support alternative spellings fairly easily in the future without breaking backward compatibility.  So maybe if there's sufficient demand, we can revisit it in a future version.